### DISCUSSION TITLE
Possible to run vim functions in key mappings?

### DISCUSSION THREAD

**Discussion by jphalip** [JetBrains/ideavim#1036] [Q&A]
Hi, I've got these key mappings in my regular Vim config:

```
" Pre-populate the search field when text is selected and you press /
vnoremap / "sy/\V<C-R>=escape(@",'/')<CR>

" If you press <Escape> during a search before you press enter, keep cursor
" at the first search result instead of jumping back to where the cursor was
" before the search started.
cnoremap <expr> <Esc> getcmdtype() =~ '[/?]' ? '<CR><Esc>' : '<Esc>'

" Remap Enter to 'n' when search highlighting is active and there's a pattern
nnoremap <expr> <CR> (&hlsearch && v:hlsearch) ? 'n' : '<CR>'
```

BTW, those mappings are to make search functionality work a bit more similar to how search works in a browser or even with IntelliJ's built-in search.

However, those mappings don't work in IdeaVim. I assume maybe because IdeaVim can't interpret Vim functions and statements like `getcmdtype()`, `&hlsearch`, or `v:hlsearch`? Are there any ways to make this work, or do you think this functionality could be added?

Thank you.

**Comment by lippfi** [comment]
Hi!

**TL;DR:** I'll be happy to answer your questions and help you with a PR, but maintainers are currently focused on more fundamental features like remote development and Fleet IDE support.

IdeaVim does support Vim script and can interpret things (it understands that `getcmdtype()` is a function, `&hlsearch` is an option, and `v:hlsearch` is a `v:` scoped variable). However, it does not implement most of Vim's built-in functions, as there are many of them and we currently have more pressing issues and bugs to address in YouTrack.
I really hope that people will create PRs for the functions they need, and one day there will be a decent set of implemented functions.

Regarding the functions you have:
`escape()` - not implemented
`getcmdtype()` - not implemented
`&hlsearch` - works (you can try `:echo &hlsearch`)
`v:hlsearch` - not implemented (I don't really know how it differs from `&hlsearch`)

You can use the sin() function as a reference for implementing your own functions: https://github.com/JetBrains/ideavim/blob/master/vim-engine/src/main/kotlin/com/maddyhome/idea/vim/vimscript/model/functions/handlers/SinFunctionHandler.kt
All you need to do is extend the `FunctionHandler()` class and annotate your function with `@VimscriptFunction` - it will be registered automatically.

**Comment by jphalip** [comment]
Hi @lippfi. Thanks for your detailed answer! I'm happy to take a stab at this. I just started looking into it and had a few questions:

* On `getcmdtype`:

I've got a basic implementation of the `getcmdtype()` function working. However, this doesn't seem to work as expected:

```
" If you press <Escape> during a search before you press enter, keep cursor
" at the first search result instead of jumping back to where the cursor was
" before the search started.
cnoremap <expr> <Esc> getcmdtype() =~ '[/?]' ? '<CR><Esc>' : '<Esc>'
```

Could it be that other parts of this mapping are not implemented? Maybe the tertiary operator? Can you spot what the issue might be?

* On `escape`:

My `escape` function implementation never seems to be called in the following scenario:

```vim
vnoremap / "sy/\V<C-R>=escape(@",'/')<CR>
```

I think it's because it looks like IdeaVim doesn't support nested command prompts. Here normally the first prompt (i.e. the search prompt) would be triggered when you press `/`, and then the second prompt (i.e. the expression prompt) would be triggered when you press `<C-R>=` from inside the search prompt. My guess is that this behavior would be quite hard to implement - what do you think?

In the meantime, I was able to get an alternative way that achieves the same thing by yanking the output of the `escape` function into a register, and then pasting it into the search prompt:

```vim
vnoremap / "sy:let @/='\V' . escape(@s, '/')<CR>/<C-R>s<CR>N
```

* On `v:hlsearch`:

According IdeaVim's code, the `v:` scope isn't yet implemented: https://github.com/JetBrains/ideavim/blob/791edbd29b6de7db64fd622744d21001a8c54d17/vim-engine/src/main/kotlin/com/maddyhome/idea/vim/vimscript/services/VimVariableServiceBase.kt#L174-L184

I could try to add an entry specifically for "hlsearch". For that I would need to access the underlying `com.intellij.openapi.editor.Editor` object and check if some text is highlighted. However, somehow I can't access the `ij` attribute of the `VimEditor` object in that function. Do you know why that is? Can you think of another way to access the `Editor` object?

Thanks in advance!

**Comment by lippfi** [comment]
> Could it be that other parts of this mapping are not implemented? Maybe the tertiary operator? Can you spot what the issue might be?

I don't think so, I've just tried:
```
cnoremap <expr> <Esc> col('.') == 1 ? '<CR><Esc>' : '<Esc>'
```
and it worked, so it's either `getcmdtype()` not working or `=~` operator not matching correctly. However, I've also checked the `=~` operator with `echo '/' =~ '[/?]'` and it was also working

> I think it's because it looks like IdeaVim doesn't support nested command prompts

You are absolutely right, I hadn't thought about that, thank you. The `<C-R>=` is not implemented in command mode and it would be difficult to implement in IdeaVim:
While the vim-engine API allows creating multiple prompts, the IdeaVim implementation reuses the same prompt repeatedly for optimization reasons. So you can't really create a second prompt while one is already active.
I'll discuss it tomorrow with IdeaVim maintainer, but fix most likely won't be implemented soon, so it's great that you've found an alternative way

> However, somehow I can't access the ij attribute of the VimEditor object in that function. Do you know why that is? 

You can see that project root has the `src` directory with IJ platform classes (IdeaVim plugin) and the `vim-engine` directory with no IJ classes at all. It's because we aim vim-engine to become an open source library that could be used in any other IDE, text editor, browser (with WASM). vim-engine is already used in Fleet which is not an IJ Platform based product.
So vim-engine is the API with basic implementations; and src is the implementation details for IJ Platform.
You have two options really:
1. Do not change the `VimVariableServiceBase.kt` class, which is common for IJ and other platforms and override the `getVimVariable` method in `IjVariableService` (IJ platform version of the class, that knows about Editor). But this option is less preffered
2. Create an interface or add a method to an existing interface. For example, you could add `isSomeTextHighlighted()` method to the `VimSearchGroup` interface or introduce your own interface `VimHighlightingService` and register it in `VimInjector`. This method will become available to vim-engine. After this, you can implement the method in the corresponding IntelliJ class, and I will implement this method in Fleet. This way, the v:hlsearch variable will work on all platforms, not only IntelliJ

**Reply by citizenmatt** [reply]
Another alternative would be to implement `v:hlsearch` as a flag that `VimVariableSearchBase` maintains (could be a `Map<String, VimDataType>` to support future variables, or even just a simple boolean as it's the first/only one, and we can refactor later). We've currently got [a similar flag in `IjVimSearchGroup`](https://github.com/JetBrains/ideavim/blob/82d18cfbb95487da85c4cd131836746bc78c2e49/src/main/java/com/maddyhome/idea/vim/newapi/IjVimSearchGroup.kt#L67) that is either the value of the `'hlsearch'` flag, or set to `false` by `:nohlsearch`.

If we move this state and let the variable service maintains it, then `IjVimSearchGroup` can just check the `'hlsearch'` option and the `v:hlsearch` variable when deciding to show highlights - if both are `true`, then show them. When `:nohlsearch` is called, set the variable instead of setting the private property. The variable is now the source of truth, so getting the value will always be correct. Setting the value will need to update search highlights. There's a method on `VimSearchGroupBase` that will need moving to `VimSearchGroup` for that.

**Reply by jphalip** [reply]
Thank you both for your replies.

I got `getcmdtype` to work. The pattern matching was failing because of a dumb mistake on my part. I'll submit a PR later on.

Re: nested prompts. In Vim you can also trigger that use case with the `:insert` and `:append` command. There's also so-called "input command" (label: `@`) but I'm not sure how that works - Do you know?

For reference, here's the manual entry:
```
getcmdtype()                                                      *getcmdtype()*
		Return the current command-line type. Possible return values
		are:
		    :	normal Ex command
		    >	debug mode command |debug-mode|
		    /	forward search command
		    ?	backward search command
		    @	|input()| command
		    `-`	|:insert| or |:append| command
		    =	|i_CTRL-R_=|
		Only works when editing the command line, thus requires use of
		|c_CTRL-\_e| or |c_CTRL-R_=| or an expression mapping.
		Returns an empty string otherwise.
		Also see |getcmdpos()|, |setcmdpos()| and |getcmdline()|.
```

Re: `hlsearch`, I got it working by adding the `isSomeTextHighlighted()` method as you suggested. I'll submit a PR for that as well.

I'll post back here once the PRs are submitted. Thanks again!

**Comment by jphalip** [comment]
Here's the first PR for the `escape()` function: https://github.com/JetBrains/ideavim/pull/1038

**Comment by jphalip** [comment]
Here's the PR for `getcmdtype()`. I'd need some guidance on how to write unit tests for it: https://github.com/JetBrains/ideavim/pull/1039

**Comment by jphalip** [comment]
And here's the PR for `v:hlsearch`: https://github.com/JetBrains/ideavim/pull/1040

BTW I've also recently a couple other PRs, in case you have a chance to review them:
https://github.com/JetBrains/ideavim/pull/1035
https://github.com/JetBrains/ideavim/pull/1031

**Comment by jphalip** [comment]
BTW, something kind of not related. I've started implementing an IntelliJ port of [vim-switch](https://github.com/AndrewRadev/switch.vim). I'm actually almost done with it, but then it occurred to me: wouldn't it be better to instead implement any missing functions/features in IdeaVim that the plugin would require?

So my question to you is: When you declare a plugin with something like: `Plug 'AndrewRadev/switch.vim'` - would IdeaVim then load that plugin and attempt to run its commands/functions/etc when prompted? Or can plugins only work if they're converted into Java/kotlin code, and then embedded in IdeaVim (like all of those under `src/main/java/com/maddyhome/idea/vim/extension`) or installed as an IntelliJ plugin from the Marketplace?

**Reply by citizenmatt** [reply]
Right now, IdeaVim doesn't support script-based plugins. You could use `source` in your `~/.ideavimrc` to load other scripts, but there's no support for runtime directories to be able to load plugins in any meaningful way. IdeaVim's scripting subsystem also doesn't support script local functions and variables (`<SID>` and `s:`), which would likely cause problems with "real" Vim scripts. It would be good to support both, one day.

But that means that currently, IdeaVim extensions are Java/Kotlin code that is either part of IdeaVim or released as a standalone IntelliJ plugin with a dependency on IdeaVim. The `Plug` syntax was really chosen for familiarity - the extension is an emulation of a known plugin, and this is a common way to enable it in both Vim and IdeaVim. Originally, IdeaVim only supported enabling extensions through a custom option, but it's a non-standard way of enabling options, so `Plug` is a good alternative to that (on the other hand, custom options allow disabling plugins, which `Plug` doesn't support...)
